File: coin-change/coin_change.clj
(defn coin-change [coins amount]
  (let [dp (vec (concat [0] (repeat amount Double/POSITIVE_INFINITY)))]
    (reduce (fn [dp coin]
             (reduce (fn [dp i]
                      (assoc dp i (min (nth dp i) (+ (nth dp (- i coin)) 1))))
                    dp (range coin (inc amount))))
           dp coins)
    (let [result (nth dp amount)]
      (if (= result Double/POSITIVE_INFINITY) -1 result))))

(println (coin-change [1 3 4] 6))

File: coin-change/coin_change.rs
fn coin_change(coins: &[i32], amount: i32) -> i32 {
    let mut dp = vec![i32::MAX; (amount + 1) as usize];
    dp[0] = 0;
    
    for &coin in coins {
        for i in coin..=amount {
            if dp[(i - coin) as usize] != i32::MAX {
                dp[i as usize] = dp[i as usize].min(dp[(i - coin) as usize] + 1);
            }
        }
    }
    
    if dp[amount as usize] == i32::MAX { -1 } else { dp[amount as usize] }
}

fn main() {
    println!("{}", coin_change(&[1, 3, 4], 6));
}